pydantic_ai_slim/pydantic_ai/ui/vercel_ai/_adapter.py

```diff
@@ -197,3 +215,226 @@ def load_messages(cls, messages: Sequence[UIMessage]) -> list[ModelMessage]:  #
                 assert_never(msg.role)
 
         return builder.messages
+
+    @classmethod
+    def dump_messages(  # noqa: C901
+        cls,
+        messages: Sequence[ModelMessage],
+        *,
+        _id_generator: Callable[[], str] | None = None,
+    ) -> list[UIMessage]:
+        """Transform Pydantic AI messages into Vercel AI messages.
+
+        Args:
+            messages: A sequence of ModelMessage objects to convert
+            _id_generator: Optional ID generator function for testing. If not provided, uses uuid.uuid4().
+
+        Returns:
+            A list of UIMessage objects in Vercel AI format
+        """
+
+        def _message_id_generator() -> str:
+            """Generate a message ID."""
+            return _id_generator() if _id_generator is not None else str(uuid.uuid4())
+
+        tool_returns: dict[str, ToolReturnPart | BuiltinToolReturnPart] = {}
+        tool_errors: dict[str, RetryPromptPart] = {}
+
+        for msg in messages:
+            if isinstance(msg, ModelRequest):
+                for part in msg.parts:
+                    if isinstance(part, ToolReturnPart | BuiltinToolReturnPart):
```

Comment:
`BuiltinToolReturnPart` only exists on `ModelResponse` as it all happens on the server side